### PR TITLE
fix(llc): Fixing watchers stream [ISSUE-1432] 

### PR DESCRIPTION
**Pull request by leandroBorgesFerreira**
# Submit a pull request
## CLA

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required).
- [x] The code changes follow best practices
- [x] Code changes are tested (add some information if not applicable)

## Description of the pull request
Listening for events `user.watching.start` and `user.watching.stop` and adding data to `ChannelState.watchersStream`

### CODE REVIEW

File: packages/stream_chat/lib/src/client/channel.dart [hunk 1]
```diff
@@ -1770,6 +1794,19 @@ class ChannelClientState {
     ));
   }
 
+  void _incrementWatcher(User user) {
+    List<User> newWatchers;
+
+    if (channelState.watchers != null) {
+      channelState.watchers!.add(user);
```
Comment (xsahil03x) [inline packages/stream_chat/lib/src/client/channel.dart:1801]: we should not update the `channelState` properties directly, rather just emit a new state with the update `channelState`.

```dart
updateChannelState(
    channelState.copyWith(
        watchers: updatedWatcher
    )
);
```

File: packages/stream_chat/lib/src/client/channel.dart [hunk 2]
```diff
@@ -1754,6 +1758,26 @@ class ChannelClientState {
...
+        }
+      }),
+    );
+  }
+
+  void _listenUserStopWatching() {
+    _subscriptions.add(
+      _channel.on(EventType.userWatchingStop).listen((event) {
+        if (event.user != null) {
+          channelState.watchers?.remove(event.user);
```
Comment (xsahil03x) [inline packages/stream_chat/lib/src/client/channel.dart:1775]: Similar thing here

**Review by xsahil03x — CHANGES_REQUESTED** [review]
Left some comments, you can also follow the logic we have for members.

```dart
void _listenMemberAdded() {
    _subscriptions.add(_channel.on(EventType.memberAdded).listen((Event e) {
      final member = e.member;
      final existingMembers = channelState.members ?? [];
      updateChannelState(channelState.copyWith(
        members: [
          ...existingMembers,
          member!,
        ],
      ));
    }));
  }
  
void _listenMemberRemoved() {
    _subscriptions.add(_channel.on(EventType.memberRemoved).listen((Event e) {
      final user = e.user;
      final existingMembers = channelState.members ?? [];
      updateChannelState(channelState.copyWith(
        members: existingMembers
            .where((m) => m.userId != user!.id)
            .toList(growable: false),
      ));
    }));
  }
```

File: packages/stream_chat/lib/src/client/channel.dart [hunk 3]
```diff
@@ -1770,6 +1808,19 @@ class ChannelClientState {
...
+
+    if (channelState.watchers != null) {
+      channelState.watchers!.add(user);
+      newWatchers = channelState.watchers!;
+    } else {
+      newWatchers = [user];
+    }
+
+    channelState.copyWith(watchers: newWatchers);
+  }
```
[the remark this answers is not in the document]
Comment (leandroBorgesFerreira) [reply packages/stream_chat/lib/src/client/channel.dart:1822]: Indeed, sorry I forgot this one.

**Review by xsahil03x — APPROVED** [review]

**Pull request merged**